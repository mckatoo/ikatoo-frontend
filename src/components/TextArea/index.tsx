import React, { TextareaHTMLAttributes, useEffect, useState } from 'react'

import Styles from './styles'

export type TextAreaProps = {
  onTextAreaChange?: (value: string) => void
  label?: string
  labelColor?: 'black' | 'white'
  initialValue?: string
  disabled?: boolean
  error?: string
  name: string
} & TextareaHTMLAttributes<HTMLTextAreaElement>

const TextArea = ({
  label,
  labelColor = 'black',
  name,
  initialValue = '',
  error,
  disabled = false,
  onTextAreaChange,
  ...props
}: TextAreaProps) => {
  const [value, setValue] = useState(initialValue)

  useEffect(() => {
    setValue(initialValue)
  }, [initialValue])

  const onChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    const newValue = e.currentTarget.value
    setValue(newValue)

    !!onTextAreaChange && onTextAreaChange(newValue)
  }

  return (
    <Styles.Wrapper disabled={disabled} error={!!error}>
      {!!label && (
        <Styles.Label labelColor={labelColor} htmlFor={name}>
          {label}
        </Styles.Label>
      )}
      <Styles.ElementWrapper>
        <Styles.Element
          onChange={onChange}
          value={value}
          disabled={disabled}
          name={name}
          {...(label ? { id: name } : {})}
          {...props}
        />
      </Styles.ElementWrapper>
      {!!error && <Styles.Error>{error}</Styles.Error>}
    </Styles.Wrapper>
  )
}

export default TextArea
