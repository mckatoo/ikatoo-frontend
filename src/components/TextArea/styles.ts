import styled, { css } from 'styled-components'
import tw from 'tailwind-styled-components'

import { TextAreaProps } from '.'
import theme from '../../theme'

type LabelProps = Pick<TextAreaProps, 'labelColor'>

type WrapperProps = Pick<TextAreaProps, 'disabled'> & { error?: boolean }

const ElementWrapper = styled.div`
  ${() => css`
    display: flex;
    align-items: center;
    background: ${theme.colors.lightGray};
    border-radius: 0.2rem;
    padding: 0 0 0 ${theme.spacings.xsmall};
    border: 0.2rem solid;
    border-color: ${theme.colors.lightGray};

    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }
  `}
`

const Element = styled.textarea`
  color: ${theme.colors.black};
  font-family: ${theme.font.family};
  font-size: ${theme.font.sizes.large};
  padding: ${theme.spacings.xsmall} 0;
  background: transparent;
  border: 0;
  outline: none;
  &:-webkit-autofill {
    -webkit-box-shadow: 0 0 0 ${theme.spacings.small} ${theme.colors.lightGray}
      inset;
    filter: none;
    &::first-line {
      font-family: ${theme.font.family};
      font-size: ${theme.font.sizes.large};
    }
  }
`

const Label = styled.label<LabelProps>`
  ${({ labelColor }) => css`
    font-size: ${theme.font.sizes.large};
    color: ${theme.colors[labelColor ?? 'black']};
    cursor: pointer;
  `}
`

const Error = styled.p`
  ${() => css`
    color: ${theme.colors.red};
    font-size: ${theme.font.sizes.xsmall};
  `}
`

const wrapperModifiers = {
  error: () => css`
    ${ElementWrapper} {
      border-color: ${theme.colors.red};
    }

    ${Label} {
      color: ${theme.colors.red};
    }
  `,
  disabled: () => css`
    ${Label},
    ${Element} {
      cursor: not-allowed;
      color: ${theme.colors.gray};

      &::placeholder {
        color: currentColor;
      }
    }
  `
}

const Wrapper = styled.div<WrapperProps>`
  ${({ error, disabled }) => css`
    ${error && wrapperModifiers.error()}
    ${disabled && wrapperModifiers.disabled()}
  `}
`

export default {
  Wrapper: tw(Wrapper)`
    max-w-full
    h-full
    max-h-full
  `,
  Error,
  Label,
  Element: tw(Element)`
    w-full
    max-h-[50vh]
    overflow-y-auto
  `,
  ElementWrapper
}
