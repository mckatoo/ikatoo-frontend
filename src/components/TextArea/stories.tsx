import { ComponentMeta, ComponentStory } from '@storybook/react'
import TextArea from '.'

export default {
  title: 'Components/TextArea',
  component: TextArea,
  args: {
    label: 'E-mail',
    name: 'email',
    initialValue: '',
    placeholder: 'john.cage@gmail.com',
    disabled: false
  }
} as ComponentMeta<typeof TextArea>

export const Default: ComponentStory<typeof TextArea> = (args) => (
  <div className="w-full p-15">
    <TextArea {...args} />
  </div>
)

export const withError: ComponentStory<typeof TextArea> = (args) => (
  <div className="w-80 p-15">
    <TextArea {...args} />
  </div>
)
withError.args = {
  error: 'Ops...something is wrong'
}

export const WhiteLabel: ComponentStory<typeof TextArea> = (args) => (
  <div className="w-80 p-15">
    <TextArea {...args} />
  </div>
)
WhiteLabel.args = {
  labelColor: 'white'
}
