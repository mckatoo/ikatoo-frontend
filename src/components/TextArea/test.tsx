import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { vi } from 'vitest'

import TextArea from '.'

describe('<TextArea />', () => {
  it('Renders with Label', () => {
    render(<TextArea label="Label" name="Label" />)

    expect(screen.getByLabelText('Label')).toBeInTheDocument()
  })

  it('Renders without Label', () => {
    render(<TextArea name="username" />)

    expect(screen.queryByLabelText('Label')).not.toBeInTheDocument()
  })

  it('Renders with placeholder', () => {
    render(<TextArea name="username" placeholder="hey you" />)

    expect(screen.getByPlaceholderText('hey you')).toBeInTheDocument()
  })

  it('Changes its value when typing', async () => {
    const onTextAreaChange = vi.fn()
    render(
      <TextArea
        onTextAreaChange={onTextAreaChange}
        label="TextArea"
        name="TextArea"
      />
    )

    const input = screen.getByRole('textbox')
    const text = 'This is my new text'
    userEvent.type(input, text)

    await waitFor(() => {
      expect(input).toHaveValue(text)
      expect(onTextAreaChange).toHaveBeenCalledTimes(text.length)
    })
    expect(onTextAreaChange).toHaveBeenCalledWith(text)
  })

  it('Does not changes its value when disabled', async () => {
    const onTextAreaChange = vi.fn()
    render(
      <TextArea
        onTextAreaChange={onTextAreaChange}
        label="TextArea"
        name="TextArea"
        disabled
      />
    )

    const input = screen.getByRole('textbox')
    expect(input).toBeDisabled()

    const text = 'This is my new text'
    userEvent.type(input, text)

    await waitFor(() => {
      expect(input).not.toHaveValue(text)
    })
    expect(onTextAreaChange).not.toHaveBeenCalled()
  })

  it('Renders with error', () => {
    const { container } = render(
      <TextArea name="username" label="TextArea" error="Error message" />
    )

    expect(screen.getByText('Error message')).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('Is accessible by tab', () => {
    render(<TextArea label="TextArea" name="TextArea" />)

    const input = screen.getByLabelText('TextArea')
    expect(document.body).toHaveFocus()

    userEvent.tab()
    expect(input).toHaveFocus()
  })

  it('Is not accessible by tab when disabled', () => {
    render(<TextArea label="TextArea" name="TextArea" disabled />)

    const input = screen.getByLabelText('TextArea')
    expect(document.body).toHaveFocus()

    userEvent.tab()
    expect(input).not.toHaveFocus()
  })
})
