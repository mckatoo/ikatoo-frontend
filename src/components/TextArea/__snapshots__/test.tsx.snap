// Vitest Snapshot v1

exports[`<TextArea /> > Renders with error 1`] = `
.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #e5e7eb;
  border-radius: 0.2rem;
  padding: 0 0 0 0.4rem;
  border: 0.2rem solid;
  border-color: #e5e7eb;
}

.c4:focus-within {
  box-shadow: 0 0 0.5rem #1e293b;
}

.c5 {
  color: #00021f;
  font-family: 'Oswald','Helvetica','Arial','sans-serif';
  font-size: 0.9rem;
  padding: 0.4rem 0;
  background: transparent;
  border: 0;
  outline: none;
}

.c5:-webkit-autofill {
  -webkit-box-shadow: 0 0 0 0.6rem #e5e7eb inset;
  -webkit-filter: none;
  filter: none;
}

.c5:-webkit-autofill::first-line {
  font-family: 'Oswald','Helvetica','Arial','sans-serif';
  font-size: 0.9rem;
}

.c2 {
  font-size: 0.9rem;
  color: #00021f;
  cursor: pointer;
}

.c6 {
  color: #ff2a2a;
  font-size: 0.6rem;
}

.c0 .c3 {
  border-color: #ff2a2a;
}

.c0 .c1 {
  color: #ff2a2a;
}

<div
  class="c0 max-w-full h-full max-h-full"
>
  <label
    class="c1 c2"
    for="username"
  >
    TextArea
  </label>
  <div
    class="c3 c4"
  >
    <textarea
      class="c5 w-full max-h-[50vh] overflow-y-auto"
      id="username"
      name="username"
    />
  </div>
  <p
    class="c6"
  >
    Error message
  </p>
</div>
`
